### PR TITLE
Support --mtime flag

### PR DESCRIPTION
**Pull request by fejta**
https://github.com/bazelbuild/bazel/commit/25d202fc1545f4aaba821df75a8bf3fb57284da1#diff-c1e1cca79dc0d65a1615802308293d12 which is in 0.24.0 started adding this flag

/assign @ixdy @mikedanese 

ref https://github.com/kubernetes/test-infra/pull/11987

### CODE REVIEW

File: tools/build_tar/buildtar.go [hunk 1]
```diff
@@ -222,17 +227,14 @@ func (f *tarFile) addFile(file, dest string) error {
...
-		Name:  dest,
-		Mode:  int64(mode),
-		Uid:   uid,
-		Gid:   gid,
-		Size:  0,
-		Uname: uname,
-		Gname: gname,
-	}
-
-	if err := f.makeDirs(header); err != nil {
```
Comment (ixdy) [inline tools/build_tar/buildtar.go]: why was this deleted?

File: tools/build_tar/buildtar.go [hunk 2]
```diff
@@ -276,6 +278,7 @@ func (f *tarFile) addLink(symlink, target string) error {
 		Name:     symlink,
 		Typeflag: tar.TypeSymlink,
 		Linkname: target,
+		ModTime:  mtime,
```
Comment (ixdy) [inline tools/build_tar/buildtar.go]: I think you can use `f.meta.mtime` if you want to avoid passing it as an argument

File: tools/build_tar/buildtar.go [hunk 3]
```diff
@@ -559,6 +567,33 @@ func (f *fileMeta) getMode(fname string) os.FileMode {
...
+}
+
+func (ts *timeString) Set(v string) error {
+	if v == portableTime {
+		*ts = timeString(y2k)
+		return nil
+	}
+	unix, err := strconv.ParseInt(v, 10, 64)
+	if err != nil {
+		return fmt.Errorf("not an integer or %s: %v", portableTime, v)
```
Comment (ixdy) [inline tools/build_tar/buildtar.go]: what if `v` is empty string (e.g. older bazel)?

File: tools/build_tar/buildtar.go (hunk 1 above)
Comment (fejta) [reply tools/build_tar/buildtar.go]: Done

File: tools/build_tar/buildtar.go (hunk 2 above)
Comment (fejta) [reply tools/build_tar/buildtar.go]: Done

File: tools/build_tar/buildtar.go (hunk 3 above)
Comment (fejta) [reply tools/build_tar/buildtar.go]: Done

**Comment by fejta** [comment]
And now with passing tests!

File: tools/BUILD.bazel [hunk 4]
```diff
@@ -22,12 +22,9 @@ exports_files(["CROSSTOOL"])
     name = "cc-gcc-" + cpu,
     all_files = ":empty",
     compiler_files = ":empty",
-    cpu = cpu,
```
Comment (ixdy) [inline tools/BUILD.bazel]: I'm fine having these changes in this PR, but can you put them in a separate commit?

File: tools/build_tar/buildtar.go [hunk 5]
```diff
@@ -559,6 +571,38 @@ func (f *fileMeta) getMode(fname string) os.FileMode {
...
+	return strconv.FormatInt(t.Unix(), 10)
+}
+
+func (ts *timeString) Set(v string) error {
+	var t time.Time
+	switch v {
+	case "":
+		// Use zero
+	case portableTime:
+		*ts = timeString(y2k)
```
Comment (ixdy) [inline tools/build_tar/buildtar.go]: I think this needs to return? otherwise won't L602 override?

Comment (ixdy) [reply tools/build_tar/buildtar.go]: alternately do `t = y2k` here?

File: tools/build_tar/buildtar.go [hunk 6]
```diff
@@ -559,6 +571,38 @@ func (f *fileMeta) getMode(fname string) os.FileMode {
...
+		return portableTime
+	}
+	return strconv.FormatInt(t.Unix(), 10)
+}
+
+func (ts *timeString) Set(v string) error {
+	var t time.Time
+	switch v {
+	case "":
+		// Use zero
```
Comment (ixdy) [inline tools/build_tar/buildtar.go]: also the zero value of time.Time is not the same as unix epoch 0, apparently?

```golang
	var t time.Time
	fmt.Printf("%s\n", t)
```
yields `0001-01-01 00:00:00 +0000 UTC`

Comment (ixdy) [reply tools/build_tar/buildtar.go]:
```golang
	var t time.Time
	fmt.Printf("%d\n", t.Unix())
```
-> `-62135596800`

who knew?

(I think unix epoch is probably a more expected "0" value?)

Comment (fejta) [reply tools/build_tar/buildtar.go]: Does not matter

Comment (ixdy) [reply tools/build_tar/buildtar.go]: it does if we're wanting to maintain existing behavior.

Comment (fejta) [reply tools/build_tar/buildtar.go]: It doesn't, and that would change the behavior

Comment (fejta) [reply tools/build_tar/buildtar.go]: And the point is to get things done so you can move onto the next thing, not spend forever asymptotically approaching a local maximum.

Comment (ixdy) [reply tools/build_tar/buildtar.go]: the current implementation of `build_tar` doesn't set the modification time, so tarballs implicitly use unix epoch.

this new implementation would've changed the default for older bazel (such as 0.23, which don't pass the `--mtime` flag) to use 0001-01-01, which is wildly different.

Comment (fejta) [reply tools/build_tar/buildtar.go]: That's now what setting `Header.ModTime = time.Time{}` does

Comment (ixdy) [reply tools/build_tar/buildtar.go]: ah, interesting. clearly I did not know that.
